std/array_list.zig

```diff
@@ -193,27 +194,97 @@ pub fn AlignedArrayList(comptime T: type, comptime A: u29) type {
             return self.pop();
         }
 
+        // an iterator which allows you to remove the current element
+        // while iterating through it.
         pub const Iterator = struct {
-            list: *const Self,
-            // how many items have we returned
-            count: usize,
+            list: *Self,
+
+            // index of the next element to return
+            next_index: usize = 0,
+
+            // the last index we removed
+            removed_index: ?usize = null,
 
             pub fn next(it: *Iterator) ?T {
-                if (it.count >= it.list.len) return null;
-                const val = it.list.at(it.count);
-                it.count += 1;
+                if (it.next_index >= it.list.len) return null;
+                const val = it.list.at(it.next_index);
+
+                it.next_index += 1;
+
                 return val;
             }
 
             pub fn reset(it: *Iterator) void {
-                it.count = 0;
+                it.next_index = 0;
+                it.removed_index = null;
+            }
+
+            //
+            // NOTE: inlining these removal routines is important.
```

Comment:
I would allow the caller to decide if they want to inline or not. Note that they may prefer smaller code size over speed. The optimiser should really be picking correctly for them.

Reply:
The slowdown exhibits in release-safe and debug mode; 8 and 5 percent respectively.

It seems rather inconvienent to have to call `_ = @inlineCall(itr.swapRemove)` for every call you want inlined if you want fast code, as you normally do.

Certainly in release-safe, you don't want to pay in speed any more than you are already.

Maybe we want a `@strongInlineHint()`, or `@setInlineHint(.Strong)` builtin; meaning, "you don't have to inline this, but you really should unless you really don't want to."
That way, you could have the best of both worlds, despite the fact that I don't really like the complication...

Or even better, IMO, something like:
```zig
comptime {
    if (builtin.mode != .ReleaseSmall) {
        // @enclosingFunction().force_inline = true;
        @typeInfo(@This().swapRemove).Fn.force_inline = true;
    }
}
```